### ISSUE TITLE
New api, onload signal decorator - wishlist

### ISSUE THREAD

**Issue by zbik2607** [odoo/odoo#5199]
That would be very helpful signal decorator. Settings, for example, dynamic domain for field would be possible.

**Comment by Robx64** [comment]
:)

**Comment by zbik2607** [comment]
https://www.odoo.com/forum/help-1/question/equivalent-javascript-onload-it-is-possible-when-form-is-loaded-75863

**Comment by funbaker** [comment]
Is there any workaround for this? I'm trying to use a dynamic domain too...

**Comment by xmo-odoo** [comment]
@funbaker  `onchange` callbacks are called during the initial loading of forms, based on field values provided by default_get. That is on creation the form view calls `default_get`, fills the fields, then calls `BaseModel.onchange` which calls the various `onchange` callbacks.

Currently all the onchange callbacks are called, but that's temporary, ideally we should only call those for which a default value was fetched.

You could also override `BaseModel.onchange` directly, though there is no official way to differentiate between a new record being created and an existing record being edited.

**Comment by funbaker** [comment]
I'm confused. For some reason, onchange is only fired while creating a new record (Button Create/Add Item)

**Comment by xmo-odoo** [comment]
I don't understand what you're saying.

**Comment by funbaker** [comment]
Found this, didnt tried it yet.
https://www.odoo.com/es_ES/forum/help-1/question/change-domain-without-using-on-change-solved-62948#answer-62959

**Comment by nicknomo** [comment]
I just ran into this issue myself.  This is crazy if there is no way to reapply the domains when editing a record.  I just made an intricate set of onchange statements for setting domain, but when I go in and edit the record, the domain is set to default.  

I've seen people try workarounds with fields_view_get, which don't work.  I have tried the above listing attempting to use a function... that doesn't work.  There has to be a solution to this, or its a serious design flaw.

**Comment by taifu** [comment]
Yep. It would be really useful. @nicknomo how did you solve this issue?

**Comment by lmignon** [comment]
@taifu https://github.com/OCA/web/pulls?utf8=%E2%9C%93&q=is%3Apr%20is%3Aopen%20web_domain_field

**Comment by taifu** [comment]
@lmignon sorry, I don't get it... Where is the onload new event there? Or the workaround?

**Comment by lmignon** [comment]
@taifu The two PRs add a new module named web_domain_field that allow you to use a regular field as value for the domain attribute into a form.  More info here https://github.com/acsone/web/blob/64db0709231a8203436926f535b085fd4246bcd5/web_domain_field/README.rst

**Comment by taifu** [comment]
@lmignon nice! Is it possible to install the two PRs as an external addon?

**Comment by lmignon** [comment]
@taifu The PRs provide the same addons but the first one for Odoo 10.0 and the second ont for Odoo 9.0
You can checkout the branch providing the addon from acsone...

**Comment by taifu** [comment]
@lmignon ty sir! :-)

**Comment by Moro-Code** [comment]
Just wanted to commend on this in case anyone is facing the same issue with Odoo 13.0.
The module here works extremely well for this use case
https://github.com/acsone/web/tree/13.0/web_domain_field

**Comment by lmignon** [comment]
For the record, the official repo for the `web_domain_field` module is into OCA https://github.com/OCA/web/tree/13.0/web_domain_field (even if this module has been developed by acsone :smirk: )
